README.md
# PremiumProxy - Fight piracy by giving legal players something more!

Old bukkit text, not exacly describing plugin:

    This is something that you see for the first time, without anycraftbukkit modifications!

**Are you tired of baning players from your offline server?**
**Are you sad that you can't show the game to your best friend because of your online server?**

**Why don't have both!** Allow access some offline players and have different permissions for online and offline players!

Plugin use some simple proxy mechanism to allow every player to game - but also gives a way to detect if player has premium acconut (online) or not!

It's desiggned for servers that want to fight piracy by restricting/disallowing offline or promoting online players.

##### What does it mean to your offline server

    You can give "cash for start" or "more/other permissions" to legal players.

##### What does it mean to your online server

    You can allow to join any player to your online server, for example only as "guest/viewer"

### How to install

- Download PremiumProxy.jar to plugins
- Set online-mode to true
- Restart server

You got server that allows anybody to join, but checks if player is online or offline

- [Option] change proxy port if it is used

### Known problem:

http://mc.dlk.pl/?p=Instrukcje [PL only]

Sumary: Offline Players can't join server with this mod without single hack. This can be done two way:
- using customised MinecraftSP.jar (that can be downloaded from our webside
- adding "fake session server" to *hosts* file

### Why do i need to download Launcher or modify hosts?

**Legal/premium/online players don't need to!**

Others needs a simple workaround (modify hosts destination) because of way how minecraft checks if you're premium. 
By default - server is checking if user is premium, then if yes, client (the game) is doing the same thing on it's own and if it's not premium - it disconnects without any message.

Using this workaround will not allow you to connect to servers that don't allow non-premium users (i.e. don't use PremiumProxy)

### For what is that port in configuration file?

You just set this port to empty/unused port on you server and forget about it!

This port is where our fake session server is running, the craftbukkit ask this port for that if player is online, and this port always responses with "YES" (that's the client-side workaround) PremiumProxy asks real server if this player is Premium after this creates event to announce it to other plugins. This port allows to not modify any craftbukkit code :)

### For developers

It is possible that you will receive event before or after player login event, so keep that in mind.

The only thing you need to do is do catch PremiumStatusEvent and do with the player what you would like to do.
Example plugin code that changes player group (using Vault) from "guest" to "player" if online:

    @EventHandler(priority = EventPriority.NORMAL)
     public void playergroupchange(me.kolorafa.premiumproxy.PremiumStatusEvent event) {
     if(event.hasPremium()){
      //Your code what to do if player is legal
     }
    }


## Known addons:

- PremiumGroupChanger - Example plugin that changes player group from "nonpremium" to "premium" if Joined player is Premium (online) user.
- Premium Money - Plugin gives money to premium players on first join, prevents players from joining multiple times with fake names to gain money ;)
- Make your own and PM me :)

## Something doesn't work? Create ticket!


src/CHANGELOG
v3.0
*  Code rewrite because mojang changed player verify method, but fully compatible with old PremiumProxy

v2.1
*  Fixed unable to find valid certification path

v2.0
*  Added proxy event
*  Added PremiumProxyAsk

v1.6
*  Added "off" state at listenHost to disable listener.

v1.5 - 23.08.2012
*  Added configurable hosts

v1.4 - 23.08.2012
*  Changed from bukkit scheduler to simple Thread

v1.3 - 13.08.2012
*  Added code to throw non-premium on ioerror - configurable

v1.2 - 13.08.2012
*  Added simple warning if server is on offline mode

v1.1 - 13.08.2012
*  Removed some wrongly added Dependence

v1.0 - 11.08.2012
*  Initial version.

src/me/kolorafa/premiumproxy/PremiumCheck.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.Proxy;
import java.net.URI;
import java.net.URL;
import java.net.URLConnection;
import java.net.URLEncoder;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author kolorafa
 */
public class PremiumCheck implements Runnable {

    premiumproxyPlugin plugin;
    String uri;
    String login;

    public PremiumCheck(premiumproxyPlugin plugin, URI uri) {
        this.plugin = plugin;
        this.uri = uri.toString();
        String login = uri.toString();
        int start = login.indexOf("?user=") + 6;
        int stop = login.indexOf("&serverId=");
        this.login = login.substring(start, stop);
    }

    public PremiumCheck(premiumproxyPlugin plugin, String uri, String login) {
        this.plugin = plugin;
        this.uri = uri;
        this.login = login;
    }

    @Override
    public void run() {
        BufferedReader bufferedreader = null;

        try {
            URL url = new URL(uri);
            URLConnection uc = url.openConnection(Proxy.NO_PROXY);
            bufferedreader = new BufferedReader(new InputStreamReader(uc.getInputStream()));
            String s1 = bufferedreader.readLine();


            if (s1.equals("YES")) {
                PremiumStatusEvent ev = new PremiumStatusEvent(login, true, s1);
                plugin.getServer().getPluginManager().callEvent(ev);
                System.out.println(login + " is PREMIUM !");
            } else {
                PremiumStatusEvent ev = new PremiumStatusEvent(login, false, s1);
                plugin.getServer().getPluginManager().callEvent(ev);
                System.out.println(login + " is NON-PREMIUM ! (" + s1 + ")");
            }
        } catch (IOException ex) {
            if (plugin.getConfig().getBoolean("nonPremiumOnIOError")) {
                PremiumStatusEvent ev = new PremiumStatusEvent(login, false, "IOError");
                plugin.getServer().getPluginManager().callEvent(ev);
                System.out.println(login + " is NON-PREMIUM ! (IOError)");
                System.out.println("IOError checking "+ login + " for premium: "+ex.getMessage());
            }else{
                System.out.println("IOError checking "+ login + " for premium: "+ex.getMessage());
            }
        } finally {
            try {
                bufferedreader.close();
            } catch (IOException ex) {
            }
        }
    }
}


src/me/kolorafa/premiumproxy/PremiumProxyAsk.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.net.Proxy;
import java.util.ArrayList;
import org.bukkit.event.Event;
import org.bukkit.event.HandlerList;

/**
 *
 * @author kolorafa
 */
public class PremiumProxyAsk extends Event {

    private static final HandlerList handlers = new HandlerList();

    @Override
    public HandlerList getHandlers() {
        return handlers;
    }

    public static HandlerList getHandlerList() {
        return handlers;
    }
    
    private String playerName;
    private String sessId;
    ArrayList<Proxy> proxyList;
    boolean allowToJoin;
    boolean disablePremiumStatusEvent;
    
    public PremiumProxyAsk(String login, ArrayList<Proxy> proxyList, boolean defaultAllow) {
        playerName = login;
        this.proxyList = proxyList;
        this.allowToJoin = defaultAllow;
        disablePremiumStatusEvent=false;
    }

    public PremiumProxyAsk(String login, String sessId, boolean defaultAllow) {
        this.playerName = login;
        this.sessId = sessId;
        this.proxyList = null;
        this.allowToJoin = defaultAllow;
        disablePremiumStatusEvent=false;
    }
    
    public void addProxyToList(Proxy p){
        proxyList.add(p);
    }

    public ArrayList<Proxy> getProxyList(){
        return proxyList;
    }

    public void setDefaultRedirectToYes(boolean allowToJoin){
        this.allowToJoin = allowToJoin;
    }

    public boolean getDefaultRedirectToYes(){
        return this.allowToJoin;
    }

    public void disablePremiumStatusEvent(boolean yes){
        this.disablePremiumStatusEvent = yes;
    }

    public String getPlayerName() {
        return playerName;
    }

}


src/me/kolorafa/premiumproxy/PremiumProxyMain.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.io.*;
import java.net.Inet4Address;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author kolorafa
 */
public class PremiumProxyMain {

    
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
            KolorafaSocketYesServer server = new KolorafaSocketYesServer("0.0.0.0", 8123);
            server.start();
//        try {
//            
//            
//            ServerSocket s = new ServerSocket(1234, 10, Inet4Address.getLocalHost());
//            
//            while(true){
//                Socket c = s.accept();
//                BufferedReader input = new BufferedReader(new InputStreamReader(c.getInputStream()));
//                BufferedWriter output = new BufferedWriter(new OutputStreamWriter(c.getOutputStream()));
//                
//            }
//        } catch (UnknownHostException ex) {
//            Logger.getLogger(PremiumProxyMain.class.getName()).log(Level.SEVERE, null, ex);
//        } catch (IOException ex) {
//            Logger.getLogger(PremiumProxyMain.class.getName()).log(Level.SEVERE, null, ex);
//        }
    }
}


src/me/kolorafa/premiumproxy/premiumYesServer.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.io.*;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.UnknownHostException;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author kolorafa
 */
public class premiumYesServer extends Thread {

    int port;
    premiumproxyPlugin plugin;
    public boolean pracuj = true;
    public ServerSocket s;
    
    premiumYesServer(premiumproxyPlugin aThis, int aInt) {
        this.port = aInt;
        plugin = aThis;
    }

    public void zamknij(){
        pracuj=false;
        try {
            join(2000);
        } catch (InterruptedException ex) {
            Logger.getLogger(premiumYesServer.class.getName()).log(Level.SEVERE, null, ex);
        }
        try {
            s.close();
        } catch (IOException ex) {
            Logger.getLogger(premiumYesServer.class.getName()).log(Level.SEVERE, null, ex);
        }
    }
    
    @Override
    public void run() {
        String yesString =
                "HTTP/1.1 200 OK\r\n"
                + "Content-Type: text/plain; charset=UTF8\r\n"
                + "Server: premiumProxy\r\n"
                + "Connection: close\r\n\r\n"
                + "YES";
        try {
            if (plugin.getConfig().getString("listenHost").equalsIgnoreCase("off")
                    || plugin.getConfig().getString("listenHost").equalsIgnoreCase("none")
                    || plugin.getConfig().getString("listenHost").equalsIgnoreCase("false")) {
                return;
            }
            s = new ServerSocket(port, 0, InetAddress.getByName(plugin.getConfig().getString("listenHost")));
            while (pracuj) {
                Socket c = s.accept();
                try {
                    BufferedReader input = new BufferedReader(new InputStreamReader(c.getInputStream()));
                    BufferedWriter output = new BufferedWriter(new OutputStreamWriter(c.getOutputStream()));
                    output.write(yesString);
                    input.readLine();
                    while (input.ready()) {
                        input.readLine();
                    }
                    output.close();
                    c.close();
                } catch (IOException ex) {
                    System.out.println(ex);
                }
            }
            s.close();
        } catch (UnknownHostException ex) {
            System.out.println(ex);
        } catch (IOException ex) {
            System.out.println(ex);
        }
    }
}


src/me/kolorafa/premiumproxy/PremiumStatusEvent.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import org.bukkit.event.Event;
import org.bukkit.event.HandlerList;

/**
 *
 * @author kolorafa
 */
public class PremiumStatusEvent extends Event {

    private static final HandlerList handlers = new HandlerList();

    @Override
    public HandlerList getHandlers() {
        return handlers;
    }

    public static HandlerList getHandlerList() {
        return handlers;
    }
    private String playerName;
    private boolean hasPremium;
    private String returned;

    public PremiumStatusEvent(String login, boolean premium, String ret) {
        playerName = login;
        hasPremium = premium;
        returned = ret;
    }

    public String getPlayerName() {
        return playerName;
    }

    public String getReturnedString() {
        return returned;
    }

    public boolean hasPremium() {
        return hasPremium;
    }
}


src/me/kolorafa/premiumproxy/premiumproxyPlugin.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.ProxySelector;
import java.net.URL;
import java.util.ArrayList;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.bukkit.entity.Player;
import org.bukkit.plugin.PluginDescriptionFile;
import org.bukkit.plugin.java.JavaPlugin;

/**
 *
 * @author kolorafa
 */
public class premiumproxyPlugin extends JavaPlugin {

    public final Logger logger = Logger.getLogger("Minecraft");
    PluginDescriptionFile pdffile;

    public void log(String text) {
        if (getConfig().getBoolean("debug")) {
            logger.log(Level.INFO, "[" + pdffile.getName() + "] DEBUG: " + text);
        }
    }
    KolorafaSocketYesServer yesserver;
    MinecraftProxySelector ps;

    @Override
    public void onDisable() {
        if (ProxySelector.getDefault() == ps) {
            ProxySelector.setDefault(ps.defsel);
        } else {
            getLogger().info("Unable to disable PS");
        }

        yesserver.zamknij();
        if (yesserver.isAlive()) {
            logger.log(Level.INFO, "[" + pdffile.getName() + "] server did not die in time.");
            yesserver.interrupt();
        }
        logger.log(Level.INFO, "[" + pdffile.getName() + "] is disabled.");
    }

    @Override
    public void onEnable() {
        loadConfiguration();
        pdffile = this.getDescription();
        ps = new MinecraftProxySelector(this, ProxySelector.getDefault(), getConfig().getInt("port"));
        ProxySelector.setDefault(ps);
        //yesserver = new premiumYesServer(this, getConfig().getInt("port"));
        yesserver = new KolorafaSocketYesServer(this, getConfig().getInt("port"));
        yesserver.start();
        //getServer().getScheduler().scheduleAsyncDelayedTask(this, yesserver);
        if (getServer().getOnlineMode() == false) {
            log("Server offline, plugin probably will not work at all !");
        }
    }

    private void loadConfiguration() {
        getConfig().options().copyDefaults(true);
        saveConfig();
    }
}


src/me/kolorafa/premiumproxy/MinecraftProxySelector.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author kolorafa
 */
public class MinecraftProxySelector extends ProxySelector {

    ProxySelector defsel = null;
    int port;
    premiumproxyPlugin plugin;

    MinecraftProxySelector(premiumproxyPlugin plug, ProxySelector def, int pint) {
        defsel = def;
        port = pint;
        plugin = plug;
    }

    public static Map<String, List<String>> getUrlParameters(String url)
            throws UnsupportedEncodingException {
        Map<String, List<String>> params = new HashMap<String, List<String>>();
        String[] urlParts = url.split("\\?");
        if (urlParts.length > 1) {
            String query = urlParts[1];
            for (String param : query.split("&")) {
                String pair[] = param.split("=");
                String key = URLDecoder.decode(pair[0], "UTF-8");
                String value = "";
                if (pair.length > 1) {
                    value = URLDecoder.decode(pair[1], "UTF-8");
                }
                List<String> values = params.get(key);
                if (values == null) {
                    values = new ArrayList<String>();
                    params.put(key, values);
                }
                values.add(value);
            }
        }
        return params;
    }

    @Override
    public java.util.List<Proxy> select(URI uri) {
        if (uri == null) {
            throw new IllegalArgumentException("URI can't be null.");
        }
        plugin.log("PremiumProxy URI check: " + uri.toASCIIString());

        if (uri.getScheme().equalsIgnoreCase("socket") && uri.getHost().equalsIgnoreCase("session.minecraft.net")) {
            try {
                Map<String, List<String>> parms = getUrlParameters(uri.toString());
                ArrayList<Proxy> l = new ArrayList<Proxy>();
                SocketAddress addr = new InetSocketAddress(plugin.getConfig().getString("connectHost"), port);
                Proxy proxy = new Proxy(Proxy.Type.SOCKS, addr);
                l.add(proxy);
                plugin.log("redirecting .....");
                return l;
            } catch (UnsupportedEncodingException ex) {
                plugin.log("url error, what the f***");
            }
        }
//        if (uri.getScheme().equalsIgnoreCase("http") && uri.getHost().equalsIgnoreCase("session.minecraft.net")) {
//            try {
//                Map<String, List<String>> parms = getUrlParameters(uri.toString());
//                ArrayList<Proxy> l = new ArrayList<Proxy>();
//
//                plugin.log("select for " + parms.get("user").get(0) + " " + parms.get("serverId").get(0));
//
//                String login = uri.toString();
//                int start = login.indexOf("?user=") + 6;
//                int stop = login.indexOf("&serverId=");
//                login = login.substring(start, stop);
//
//                PremiumProxyAsk ev = new PremiumProxyAsk(login, l, plugin.getConfig().getBoolean("defaultRedirect", true));
//                plugin.getServer().getPluginManager().callEvent(ev);
//                if (ev.getDefaultRedirectToYes()) {
//                    SocketAddress addr = new InetSocketAddress(plugin.getConfig().getString("connectHost"), port);
//                    Proxy proxy = new Proxy(Proxy.Type.HTTP, addr);
//                    l.add(proxy);
//                    plugin.log("redirecting to YES for " + login);
//                }
//                if (!ev.disablePremiumStatusEvent) {
//                    plugin.getServer().getScheduler().scheduleAsyncDelayedTask(plugin, new PremiumCheck(plugin, uri));
//                    plugin.log("external check for " + login);
//                }
//                if (!l.isEmpty()) {
//                    return l;
//                }
//            } catch (UnsupportedEncodingException ex) {
//                plugin.log("url error, what the f***");
//            }
//        }
        if (defsel != null) {
            return defsel.select(uri);
        } else {
            ArrayList<Proxy> l = new ArrayList<Proxy>();
            l.add(Proxy.NO_PROXY);
            return l;
        }
    }

    @Override
    public void connectFailed(URI uri, SocketAddress sa, IOException ioe) {
        if (uri == null || sa == null || ioe == null) {
            throw new IllegalArgumentException("Arguments can't be null.");
        }
        defsel.connectFailed(uri, sa, ioe);
    }
}

src/me/kolorafa/premiumproxy/KolorafaSocketYesServer.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package me.kolorafa.premiumproxy;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.UnsupportedEncodingException;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.net.URLDecoder;
import java.net.UnknownHostException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author kolorafa
 */
public class KolorafaSocketYesServer extends Thread {

    int port;
    premiumproxyPlugin plugin;
    String host;
    public boolean pracuj = true;
    public ServerSocket s;

    KolorafaSocketYesServer(premiumproxyPlugin aThis, int aInt) {
        this.port = aInt;
        plugin = aThis;
        this.host = plugin.getConfig().getString("listenHost");
    }

    KolorafaSocketYesServer(String host, int port) {
        this.port = port;
        this.host = host;
    }

    public void zamknij() {
        pracuj = false;
        try {
            join(2000);
        } catch (InterruptedException ex) {
            Logger.getLogger(premiumYesServer.class.getName()).log(Level.SEVERE, null, ex);
        }
        try {
            s.close();
        } catch (IOException ex) {
            Logger.getLogger(premiumYesServer.class.getName()).log(Level.SEVERE, null, ex);
        }
    }

    @Override
    public void run() {
        String yesString =
                "HTTP/1.1 200 OK\r\n"
                + "Content-Type: text/plain; charset=UTF8\r\n"
                + "Server: premiumProxy\r\n"
                + "Connection: close\r\n\r\n"
                + "YES";

        String socketYesString = "\\x05\\x00\\x05\\x08\\x00\\x01" + yesString;

        try {
            if (host.equalsIgnoreCase("off")
                    || host.equalsIgnoreCase("none")
                    || host.equalsIgnoreCase("false")) {
                return;
            }
            s = new ServerSocket(port, 0, InetAddress.getByName(host));
            while (pracuj) {
                Socket c = s.accept();
                try {
                    BufferedReader input;
                    InputStream is = c.getInputStream();
                    BufferedWriter output = new BufferedWriter(new OutputStreamWriter(c.getOutputStream()));
                    OutputStream os = c.getOutputStream();
                    try {
                        plugin.log("new connection");
                        c.setSoTimeout(10000);
                        if (is.read() != 5) {
                            byte[] x = {0, (byte) 0xFF};
                            os.write(x);
                            throw new SocketTimeoutException("Wrong version");
                        }
                        is.read();
                        int ile;
                        boolean notfound = false;
                        while (is.available() > 0) {
                            ile = is.read();
                            if (ile == 1) {
                                notfound = true;
                            }
                        }
                        if (notfound) {
                            byte[] x = {0, (byte) 0xFF};
                            os.write(x);
                            throw new SocketTimeoutException("without login not found");
                        }
                        os.write(5);
                        os.write(0);
                        os.flush();

                        plugin.log("read command");
                        ile = is.read();
                        if (ile != 5) {
                            byte[] x = {0, 1};
                            os.write(x);
                            plugin.log("Dostalismy: " + String.valueOf(ile));
                            throw new SocketTimeoutException("wtf1");
                        }
                        if (is.read() != 1) {
                            byte[] x = {5, 7};
                            os.write(x);
                            throw new SocketTimeoutException("command not supported");
                        }
                        is.read();
                        if (is.read() != 1) {
                            byte[] x = {5, 8};
                            os.write(x);
                            throw new SocketTimeoutException("address type not supported");
                        }


                        byte ip[] = new byte[6];
                        is.read(ip, 0, 6);

                        os.write(5);
                        os.write(0);
                        os.write(0);
                        os.write(1);
                        os.write(ip, 0, 6);
                        os.flush();
                        plugin.log("Read first line");
                        input = new BufferedReader(new InputStreamReader(is));

                        String pierwszalinia = input.readLine();
                        int from = pierwszalinia.indexOf(" ");
                        int to = pierwszalinia.indexOf(" ", from + 1);
                        if (from == 0 || to == 0) {
                            plugin.log("something wrong with url");
                        }
                        plugin.log(pierwszalinia.substring(from+1, to));
                        Map<String, List<String>> parms = getUrlParameters("http://x" + pierwszalinia.substring(from+1, to));
                        plugin.log("select for " + parms.get("user").get(0) + " " + parms.get("serverId").get(0));
                        while (input.ready()) {
                            input.readLine();
                        }
                        String login = parms.get("user").get(0);
                        PremiumProxyAsk ev = new PremiumProxyAsk(login, parms.get("serverId").get(0), plugin.getConfig().getBoolean("defaultRedirect", true));
                        plugin.getServer().getPluginManager().callEvent(ev);
                        if (ev.getDefaultRedirectToYes()) {
                            plugin.log("sending YES for " + login);
                        }
                        if (!ev.disablePremiumStatusEvent) {
                            InetAddress address = InetAddress.getByName("session.minecraft.net"); 
                            plugin.getServer().getScheduler().scheduleAsyncDelayedTask(plugin, new PremiumCheck(plugin,"http://"+address.getHostAddress()+pierwszalinia.substring(from+1, to)+"&direct=kolorafa", login));
                            plugin.log("external check for " + login);
                        }

                        plugin.log("Write yes string start");
                        output.write(yesString);
                        //output.write(String.valueOf(new Random().nextInt()));
                        output.flush();
                        plugin.log("out close");
                        output.close();
                        c.close();
                    } catch (SocketTimeoutException ex) {
                        plugin.log("Timeout: " + ex.getMessage());
                        output.flush();
                        output.close();
                        c.close();
                    }
                } catch (IOException ex) {
                    plugin.log(ex.getMessage());
                }
            }
            s.close();
        } catch (UnknownHostException ex) {
            plugin.log(ex.getMessage());
        } catch (IOException ex) {
            plugin.log(ex.getMessage());
        }
    }

    public static Map<String, List<String>> getUrlParameters(String url)
            throws UnsupportedEncodingException {
        Map<String, List<String>> params = new HashMap<String, List<String>>();
        String[] urlParts = url.split("\\?");
        if (urlParts.length > 1) {
            String query = urlParts[1];
            for (String param : query.split("&")) {
                String pair[] = param.split("=");
                String key = URLDecoder.decode(pair[0], "UTF-8");
                String value = "";
                if (pair.length > 1) {
                    value = URLDecoder.decode(pair[1], "UTF-8");
                }
                List<String> values = params.get(key);
                if (values == null) {
                    values = new ArrayList<String>();
                    params.put(key, values);
                }
                values.add(value);
            }
        }
        return params;
    }
}
